homeassistant/components/tradfri/base_class.py

```diff
@@ -44,103 +38,54 @@ async def wrapper(command: Command | list[Command]) -> None:
+
         self._gateway_id = gateway_id
 
-    async def _async_run_observe(self, cmd: Command) -> None:
-        """Run observe in a coroutine."""
-        try:
-            await self._api(cmd)
-        except PytradfriError as err:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed, trying again", exc_info=err)
-            self._async_start_observe()
+        self._device: Device = device_coordinator.data
 
-    @callback
-    def _async_start_observe(self, exc: Exception | None = None) -> None:
-        """Start observation of device."""
-        if exc:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed for %s", self._attr_name, exc_info=exc)
-        cmd = self._device.observe(
-            callback=self._observe_update,
-            err_callback=self._async_start_observe,
-            duration=0,
-        )
-        self.hass.async_create_task(self._async_run_observe(cmd))
+        self._device_id = self._device.id
+        self._api = handle_error(api)
+        self._attr_name = self._device.name
 
-    async def async_added_to_hass(self) -> None:
-        """Start thread when added to hass."""
-        self._async_start_observe()
+        self._attr_unique_id = f"{self._gateway_id}-{self._device.id}"
 
+    @abstractmethod
     @callback
-    def _observe_update(self, device: Device) -> None:
-        """Receive new state data for this device."""
-        self._refresh(device)
+    def _refresh(self) -> None:
+        """Refresh device data."""
 
-    def _refresh(self, device: Device, write_ha: bool = True) -> None:
-        """Refresh the device data."""
-        self._device = device
-        self._attr_name = device.name
-        if write_ha:
-            self.async_write_ha_state()
-
-
-class TradfriBaseDevice(TradfriBaseClass):
-    """Base class for a TRADFRI device.
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """
+        Handle updated data from the coordinator.
 
-    All devices should inherit from this class.
-    """
-
-    def __init__(
-        self,
-        device: Device,
-        api: Callable[[Command | list[Command]], Any],
-        gateway_id: str,
-    ) -> None:
-        """Initialize a device."""
-        self._attr_available = device.reachable
-        self._hub_available = True
-        super().__init__(device, api, gateway_id)
+        Tests fails without this method.
+        """
+        self._refresh()
+        super()._handle_coordinator_update()
 
     async def async_added_to_hass(self) -> None:
         """Start thread when added to hass."""
-        # Only devices shall receive SIGNAL_GW
-        self.async_on_remove(
-            async_dispatcher_connect(self.hass, SIGNAL_GW, self.set_hub_available)
+        self.async_on_remove(  # Only devices shall receive SIGNAL_GW
+            async_dispatcher_connect(
```

Comment:
Please connect the gateway available callback in the coordinator instead. We only need one signal per coordinator. Maybe connect the signal in the init module when iterating over the devices and groups, after the first refresh of the coordinator? We can use `entry.async_on_unload` to make sure that the signal is disconnected on entry unload.